### ISSUE TITLE
Score2Perf: A Question Regarding Mean-Aggregation and Decoder Inputs

### ISSUE THREAD

**Issue by apteryxlabs** [magenta/magenta#1879]
Hi all,

Thank you to the Magenta team for building this - it's truly amazing to see what a community like this can accomplish. Makes 2020 that much better 👍  Full disclosure - I'm coming at this from PyTorch and have the majority of my experience working with Huggingface's models. So forgive me if I mis-parse or don't understand Tensorflow conventions yet. Go easy on me!

**Context:**
In the accompanying paper to the score2perf model, [Encoding Musical Style with Transformer Autoencoders](https://arxiv.org/abs/1912.05537), you discuss 'temporal compression' (page 2 diagram, below) - I am uncertain of how the compressed representation is attended to in the decoder:
<img width="604" alt="tfae" src="https://user-images.githubusercontent.com/65966807/104776883-c936b680-5748-11eb-8602-ebc3017c94b5.png">

The point in the code where it looks like this is implemented is here (line 110ish, transformer_autoencoder.py):
```
  if not baseline:
    encoder_output = tf.math.reduce_mean(
        encoder_output, axis=1, keep_dims=True)
    encoder_decoder_attention_bias = tf.math.reduce_mean(
        encoder_decoder_attention_bias, axis=-1, keep_dims=True)
```

**My Question(s)**
Firstly - doesn't reduce_mean include in its calculation outputs corresponding to <pad> tokens? E.g. a sequence input to the model with seq_length 123 with a model block size/max input size of 512 would have 512-123 = 389 pad tokens at its end, which (I believe) means that the final 389 vectors in the output space (of shape [1, 512, d_model]) would be meaningless (since the attention mask at input would be zero for all those positions). Shouldn't we aggregate only non-pad-related outputs?

Second - how is the resulting output vector attended to in the decode step? I'm used to encoder outputs being size [batch, block_size, d_model], not [batch, 1, d_model] (which I believe would be the case here). Are all the cross attentions attending to a single vector? Is that okay? 

Third - what do the decoder teacher-forcing inputs look like at train time? Say we're encoding input tokens [[1,2,3]]. Are our decoder inputs [[0, 1, 2]] (standard right_shifting per T5/etc causal language modeling), or are they something else? Not sure how the aggregated encoding changes things.

Any help would be greatly appreciated!

**Comment by apteryxlabs** [comment]
Bumping this - any thoughts? @cghawthorne I'm particularly stuck on what the decoder inputs should look like, and the shape of the aggregated vector. 

Correct me if I'm wrong - again, I'm coming at this from a Huggingface background -  but it looks like at **train time** you'd be providing the encoder _and_ the decoder the same inputs, possibly (but I'm not sure, looking at the code) right-shifting the decoder inputs by one, and doing a single forward pass through the decoder. But at **generation time**, encoder inputs would be the original sequence (same at train), and decoder inputs would be the <pad> token - you'd then loop the decoder to generate a new sequence. My question - how does training with a full sequence as input for the decoder translate well, or at all, to generating a new sequence from <pad>? Or am I complete missing something here?

**Comment by cghawthorne** [comment]
Good questions! @kristychoi would probably be the best to answer here.

**Comment by apteryxlabs** [comment]
Thanks @cghawthorne for your response! @kristychoi (looks like I can't tag her?), I would love to get your input. 

Our current approach is to mask out (set to zero) all encoder outputs which correspond to pad tokens, and then (rather than averaging) to stack the tensor along the seq-aka-time (hereinafter 'time') axis (where the encoder output is of shape (batch, time, d_model), and project the resulting (batch, time*d_model) tensor through a feed forward network onto a (batch, d_encoding) space. This is our autoencoded vector, which we use for later sampling.

That vector is then mapped back to the original encoder shape via another feed-forward net + reshape (in pytorch, tensor.view()) step, and the result is used for decoder attentions. The rest of the architecture is identical to yours.

Would love to hear your thoughts. Currently the decoder outputs, while containing some of the original signal, are QUITE noisy. If I'm missing anything here, I'd love to know.

I can share our repo with either of you if you're interested.

**Comment by kristychoi** [comment]
hi @apteryxlabs! let me know if anything is unclear or if i missed anything important.

- that's a good point re: potentially mean-aggregating over padded tokens. i wasn't *too* worried about this because the performance sequences were all quite long (and we had to truncate them during the data generation process), but it's possible that some were padded. the "FFN" encoding method you just described is something i'd wanted to try!
- during training, the decoder is provided two things: (1) the performance and/or melody vector representation, which was mean-aggregated across time, and (2) a perturbed performance sequence. you can think of (2) as the input with some added noise (in the music domain it was pitch shifts/time stretches, but I imagine for NLP tasks they could look like masking tokens or random word substitutions). i didn't modify anything else in terms of attention mechanisms in the decoder from what was originally there in [tensor2tensor.](https://github.com/tensorflow/tensor2tensor/blob/21dba2c1bdcc7ab582a2bfd8c0885c217963bb4f/tensor2tensor/models/transformer.py#L117)
- not all the samples we obtained at test time sounded great -- we had to filter some out. the "noisy training" helped quite a bit on this front.

**Comment by apteryxlabs** [comment]
Thanks @kristychoi! We can pretty readily expand our dataset; I'll try that plus the masking (if present). For the perturbations, masks and substitutions we'll definitely try. If we try anything else that works, I'll be sure to post it here.

Thank you for your help! It's good to know we're more or less on the right track.

**Comment by apteryxlabs** [comment]
@eeelnico You'll want to read this.
